cutils.c

```diff
@@ -674,4 +677,215 @@ int64_t js__gettimeofday_us(void) {
     return ((int64_t)tv.tv_sec * 1000000) + tv.tv_usec;
 }
 
+/* Cross-platform threading APIs. */
+
+#if defined(_WIN32)
+typedef void (*js__once_cb)(void);
+
+typedef struct {
+    js__once_cb callback;
+} js__once_data_t;
+
+static BOOL WINAPI js__once_inner(INIT_ONCE *once, void *param, void **context) {
+    js__once_data_t *data = param;
+
+    data->callback();
+
+    return TRUE;
+}
+
+void js_once(js_once_t* guard, js__once_cb callback) {
+    js__once_data_t data = { .callback = callback };
+    InitOnceExecuteOnce(guard, js__once_inner, (void*) &data, NULL);
+}
+
+void js_mutex_init(js_mutex_t* mutex) {
+    InitializeCriticalSection(mutex);
+}
+
+void js_mutex_destroy(js_mutex_t* mutex) {
+    DeleteCriticalSection(mutex);
+}
+
+void js_mutex_lock(js_mutex_t* mutex) {
+    EnterCriticalSection(mutex);
+}
+
+void js_mutex_unlock(js_mutex_t* mutex) {
+    LeaveCriticalSection(mutex);
+}
+
+void js_cond_init(js_cond_t* cond) {
+    InitializeConditionVariable(cond);
+}
+
+void js_cond_destroy(js_cond_t* cond) {
+  /* nothing to do */
+  (void) cond;
+}
+
+void js_cond_signal(js_cond_t* cond) {
+    WakeConditionVariable(cond);
+}
+
+void js_cond_broadcast(js_cond_t* cond) {
+    WakeAllConditionVariable(cond);
+}
+
+void js_cond_wait(js_cond_t* cond, js_mutex_t* mutex) {
+    if (!SleepConditionVariableCS(cond, mutex, INFINITE))
+        abort();
+}
+
+int js_cond_timedwait(js_cond_t* cond, js_mutex_t* mutex, uint64_t timeout) {
+    if (SleepConditionVariableCS(cond, mutex, (DWORD)(timeout / 1e6)))
+        return 0;
+    if (GetLastError() != ERROR_TIMEOUT)
+        abort();
+    return -1;
+}
+
+#else /* !defined(_WIN32) */
```

Comment:
```suggestion
#else /* !defined(_WIN32) */

```
For legibility. I overlooked it the first time around.